core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java

```diff
@@ -196,7 +200,7 @@
     static List<AnnotatedValueResolver> ofBeanConstructorOrMethod(Executable constructorOrMethod,
                                                                   Set<String> pathParams,
                                                                   List<RequestObjectResolver> objectResolvers) {
-        return of(constructorOrMethod, pathParams, objectResolvers, false, false, false);
+        return of(constructorOrMethod, pathParams, objectResolvers, false, false, false, null);
```

Comment:
Wondering if we want to limit this feature to only direct usage of `@Param`/`@Header` only in the service method.

i.e. The following behavior would be inconsistent
```
static class Simple {
    @Param
    public List<String> l;
}

@RegisterExtension
static final ServerExtension server = new ServerExtension() {
    @Override
    protected void configure(ServerBuilder sb) throws Exception {
        sb.annotatedService()
          .pathPrefix("/")
          .useQueryDelimiter(",")
          .build(new Object() {

            @Get("/requestObject")
            public String requestObject(@RequestObject Simple simple) {
                return Joiner.on(':').join(simple.l);
            }

            @Get("/justParam")
            public String justParam(@Param List<String> l) {
                return Joiner.on(':').join(l);
            }
        });
    }
};

@Test
void test() {
    AggregatedHttpResponse response = server.webClient().prepare().get("/requestObject")
                                            .queryParam("l", "a,b,c")
                                            .execute().aggregate().join();
    System.out.println(response.contentUtf8());

    response = server.webClient().prepare().get("/justParam")
                     .queryParam("l", "a,b,c")
                     .execute().aggregate().join();
    System.out.println(response.contentUtf8());
}
```

Reply:
This is a good point and I feel like a user may need more customization than what we have in this PR. For example, a user may want to use the delimiter only for query string and not for headers, or only for a certain query parameter/header.

For example:

```java
sb.annotatedService()
  .queryParamDelimiter(",") // For all query params
  .queryParamDelimiter("foo", ",") // Only for "foo"
  .queryParamDelimiter(queryParamName -> ",") // dynamically
  .headerDelimiter(...)
  ...
```

Reply:
At first, I imagined a '@Delimiter' annotation which represents what query delimiter to use. In this case, a user can specify the annotation with the specific '@Param' or '@Header' to enable the query delimitor to the specific param.

'''java
public void foo(@Header @Delimiter(",") List<String> delimited, @Header List<List> notDelimited);
'''

HDYT?

Reply:
> ```
> public void foo(@Header @Delimiter(",") List<String> delimited,
>                 @Header List<List> notDelimited);
> ```

Sounds good to me. :+1:

Reply:
Added `Delimiter` annotation, it has precedence over annotated service settings. And removed the flag.

Reply:
> Wondering if we want to limit this feature to only direct usage of @Param/@Header only in the service method.

> For example, a user may want to use the delimiter only for query string and not for headers, or only for a certain query parameter/header.

Would it be better to support delimiter for all of `@Param`, `@Header`, and `@RequestObject`? Currently, it's only available for the `@Param`.

Reply:
> Would it be better to support delimiter for all of `@Param`, `@Header`, and `@RequestObject`? Currently, it's only available for the `@Param`.

Yeah, it's be nice if possible. How hard would it be? How would you support it for `@RequestObject` by the way?

Reply:
> it has precedence over annotated service settings. And removed the flag.

I still think the flag is valuable. Because some users don't want to specify `@Delimeter` for every method. It may force users to do the same thing repeatedly.
They need to set the default allowed delimiters globally. For example: `-D...allowCommaDelimitedQueryParameters=&,",",;`.

Reply:
Otherwise, we can allow to configure the default delimiters with `ServerBuilder` like Trustin said.

Reply:
> I still think the flag is valuable. Because some users don't want to specify `@Delimeter` for every method.

I agree that there should be a way to specify the delimiter for a wider scope, such as on a per-server basis. However, I'm not sure if it's worth introducing a flag for this.

> Otherwise, we can allow to configure the default delimiters with `ServerBuilder` like Trustin said. 

+1

Reply:
If we introduce a method to `ServerBuilder`, may users be confused? `Server.builder().queryDelimiter(",")` isn't clear that it looks applicable for both `HttpService` and `AnnotatedService`. Or did you imagine such as `annotatedServiceQueryDelimiter()`?